### ISSUE TITLE
Not getting latest insiders version, if older insiders version downloaded in the past

### ISSUE THREAD

**Issue by gkalpak** [microsoft/vscode-test#25]
If `version` is `'insiders'` and an insiders version has been downloaded before, then the `vscode-insiders` directory will be present and downloading will be skipped:

https://github.com/microsoft/vscode-test/blob/669ba6db18fbd36474d2a9ff520ab71d542b25c5/lib/download.ts#L139

This means that insiders will never be updated once you have downloaded it the first time, unless you go and manually remove the download directory.

**Comment by octref** [comment]
I intentionally left it that way, because I thought on CI machines you start from a clean environment, and locally you wouldn't want to download it each time. I remain to be convinced though.

**Comment by gkalpak** [comment]
Ha ha! 😃
Sorry, I am not too incentivised to convince you (since I don't test against insiders and I know how to work around this issue). I was just looking through the code and noticed the issue, so I thought I'd report it.

##
Anyway, here my thoughts in case it helps:

I understand there is a tradeoff between avoiding unnecessary downloads and testing against an up-to-date version. Ideally, `vscode-test` should be able to detect whether an existing insiders download is for an outdated insiders version and re-download only if necessary.

My guess is that most people will be testing against stable versions, so if someone explicitly specifies `insiders`, then I consider it more probable that they care more about testing against the latest insiders version (than avoiding re-downloads).

As far as CI is concerned, it is not uncommon for people to cache certain directories on CI to speed up their builds ([example][1]). So, happily running your tests against a one-year old insiders download, while believing your are testing against the very latest version, won't be much fun :grin:

If there is no way to tell insiders versions apart, there are of course compromises you can make that would improve the situation (without dramatically changing the current behavior). For example, you could re-download `insiders` iff it is older than X days.

In any case, I believe this caveat is at least worth documenting, so people (a) are aware of it and (b) could manually work around it (e.g. by deleting the `.vscode-test/vscode-insiders/` directory before running their tests).

[1]: https://github.com/gkalpak/aio-docs-utils/blob/master/.travis.yml#L14-L17

**Comment by octref** [comment]
> If there is no way to tell insiders versions apart

TBH that's my main reason for leaving it this way. But I guess I can do:
- Run `code-insiders --version`  if `.vscode-test/vscode-insiders` exist
- Compare the hash with latest insiders hash

@joaomoreno https://update.code.visualstudio.com/api/releases/insider doesn't return insider hashes. 1.37.0-insider could refer to many different versions. Is there any API to get the hash?

**Comment by joaomoreno** [comment]
@octref Unfortunately not... only to download the actual bits. Would you like an API for that?

**Comment by octref** [comment]
I amended the JSDocs instead. Current behavior is consistent between stable releases and insiders (not downloading if `.vscode-test/vscode-<VERSION>` exists). Plus, if you want the said behavior, you can still do the removal yourself. It would be much harder if `vscode-test` does this automatically and you want to stop it from doing so.

**Comment by octref** [comment]
Talked to @roblourens and decided to reopen this.

**Comment by joaomoreno** [comment]
@octref Here you go:

```
https://update.code.visualstudio.com/api/versions/$VERSION/$PLATFORM/$QUALITY
```

Example:

```
https://update.code.visualstudio.com/api/versions/1.36.0-insider/darwin/insider
```

**Comment by octref** [comment]
@joaomoreno Thanks! But do we have to supplement version number such as `1.36.0-insider`? Can it be `latest-insider`?

**Comment by joaomoreno** [comment]
We've always had that for latest:

```http
GEt https://update.code.visualstudio.com/api/update/darwin/insider/latest
HTTP/1.1 200 OK

{
    "hash": "bd05eedd363dc5b823e6cda593666342cbb7fa79",
    "name": "1.37.0-insider",
    "productVersion": "1.37.0-insider",
    "sha256hash": "a866e987831acc6aacfa85daca8769b0cfeb8092c6d85345a2f2932c9c5ddee0",
    "timestamp": 1565039875561,
    "url": "https://az764295.vo.msecnd.net/insider/5e08ee01b7daee99cdda2456f19d77e715fb8762/VSCode-darwin-insider.zip",
    "version": "5e08ee01b7daee99cdda2456f19d77e715fb8762"
}
```

**Comment by octref** [comment]
Available in 1.2.0 now. @roblourens fyi.
